doc/dev/msgr2.rst

```diff
@@ -0,0 +1,199 @@
+msgr2 protocol
+==============
+
+This is a revision of the legacy Ceph on-wire protocol that was
+implemented by the SimpleMessenger.  It addresses performance and
+security issues.
+
+Definitions
+-----------
+
+* *client* (C): the party initiating a (TCP) connection
+* *server* (S): the party accepting a (TCP) connection
+* *connection*: an instance of a (TCP) connection between two peers
+* *session*: a stateful session between two peers in which message
+  exchange is ordered and lossless.  A session might span multiple
+  connections if there is an interruption (TCP connection disconnect).
+* *frame*: a discrete message sent between the peers.  Each frame
+  consists of a tag (type code), payload, and (if signing or
+  encryption is enabled) some other fields.  See below for the
+  structure.
+* *tag*: a single-byte type code associated with a frame.  The tag
+  determines the structure of the payload.
+
+Phases
+------
+
+A connection has four distinct phases:
+
+#. banner
+#. authentication
+#. message flow handshake
+#. message exchange
+
+Banner
+------
+
+Both the client and server, upon connecting, send a banner::
+
+  "ceph %x %x\n", protocol_features_suppored, protocol_features_required
+
+The protocol features are a new, distinct namespace.  Initially no
+features are defined or required, so this will be "ceph 0 0\n".
+
+If the remote party advertises features we don't support, we can
+disconnect.
+
+Authentication
+--------------
+
+A series of frames between client and server of the form::
+
+  tag byte (TAG_*)
+  payload
+
+where the payload is determiend by the tag.
+
+* TAG_AUTH_METHODS (server only): list authentication methods (none, cephx, ...)::
+
+    __le32 num_methods;
+    __le32 methods[num_methods];   // CEPH_AUTH_{NONE, CEPHX}
+
+* TAG_AUTH_SET_METHOD (client only): set auth method for this connection::
+
+    __le32 method;
+
+  - The selected auth method determines the sig_size and block_size in any
+    subsequent messages (TAG_AUTH_DONE and non-auth messages).
+
+* TAG_AUTH_BAD_METHOD (server only): reject client-selected auth method::
+
+    __le32 method
+
+* TAG_AUTH: client->server or server->client auth message::
+
+    __le32 len;
+    method specific payload
+
+* TAG_AUTH_DONE::
+    
+    __le64 flags
+      FLAG_ENCRYPTED  1
+      FLAG_SIGNED     2
+    signature
+
+  - The client first says AUTH_DONE, and the server replies to
+    acknowledge it.
+
+
+Message frame format
+--------------------
+
+Each frame can take one of two forms.  If FLAG_SIGNED or
+FLAG_ENCRYPTED has been specified::
+
+  confounder (block_size bytes of random garbage)
+  __le32 length
+  tag byte
+  payload
+  signature (sig_size bytes)
+  more confounder padding (to pad data from start of __le32 length out to block size)
```

Comment:
We just want to fill out to a block_size boundary so that we can complete the encryption block and push data over the wire.  We can't stick those extra bytes at the beginning easily because length goes there.  Unless we put len before the confounder.

Hmm, I'm not sure how useful the confounder is on *every* if we are encrypted.  Isn't one at the start sufficient?  In that case, we'd have 3 frame formats.. for unprotected, signed, and encrypted.